sklearn/datasets/samples_generator.py

```diff
@@ -735,35 +739,63 @@ def make_blobs(n_samples=100, n_features=2, centers=3, cluster_std=1.0,
     """
     generator = check_random_state(random_state)
 
-    if isinstance(centers, numbers.Integral):
-        centers = generator.uniform(center_box[0], center_box[1],
-                                    size=(centers, n_features))
+    if isinstance(n_samples, int):
+        # n_centers is figured by the centers arg
+        if isinstance(centers, numbers.Integral):
+            centers = generator.uniform(center_box[0], center_box[1],
+                                        size=(centers, n_features))
+        else:
+            centers = check_array(centers)
+            n_features = centers.shape[1]
+
+        n_centers = centers.shape[0]
+
     else:
-        centers = check_array(centers)
-        n_features = centers.shape[1]
+        # n_centers is figured by the [n_samples] arg
+        n_centers = len(n_samples)
+        if isinstance(centers, int):
```

Comment:
Better, should change centers to default to None and interpret this as 3 in the `n_samples` int case. Then raise a ValueError in the `n_samples` list case if centers is not either a 2d-array with `len(centers) == len(n_samples)` or None.